implementations/rust/ockam/ockam/src/pipe2/mod.rs

```diff
@@ -0,0 +1,159 @@
+    /// worker system pipeline.
+    fin: Address,
+}
+
+impl PipeBuilder {
+    fn new(mode: Mode) -> Self {
+        Self {
+            send_hooks: SystemBuilder::new(),
+            recv_hooks: SystemBuilder::new(),
+            peer: None,
+            recv: None,
+            fin: Address::random(0),
+            mode,
+        }
+    }
+
+    /// Construct a fixed pipe to a specific well-known peer
+    pub fn fixed() -> Self {
+        Self::new(Mode::Static)
+    }
+
+    /// Construct a pipe using dynamic initialisation handshakes
+    pub fn dynamic() -> Self {
+        Self::new(Mode::Dynamic)
+    }
+
+    /// Set a connection peer, creating an outgoing pipe
+    ///
+    /// * In `fixed` mode this attempts to connect directly to a
+    /// receiver at the given peer route.
+    ///
+    /// * In `dynamic` mode this initiates a handshake with the
+    /// given peer.  This handshake then resolves to the final
+    /// receiver
+    pub fn connect<R: Into<Route>>(mut self, peer: R) -> Self {
+        self.peer = Some(peer.into());
+        self
+    }
+
+    /// Set a receiving address, creating a receiving pipe
+    ///
+    /// * In `fixed` mode this creates a pipe receiver which waits
+    /// for incoming messages from a sender.
+    ///
+    /// * In `dynamic` mode this spawns a handshake listener, which
+    /// will create pipe receivers dynamically for any incoming
+    /// initialisation request
+    pub fn receive<A: Into<Address>>(mut self, addr: A) -> Self {
+        self.recv = Some(addr.into());
+        self
+    }
+
+    /// Set this pipe to enforce the ordering of incoming messages
+    pub fn enforce_ordering(mut self) -> Self {
+        // A pipe can currently only have two types of hooks.
+        // "Delivery" and "Ordering".  Because we want "ordering" to
+        // be applied _before_ "delivery" we need to make sure that if
+        // "delivery" has already been added, we point towards _its_
+        // address, instead of "fin".
+        let next_recv_addr = self
+            .recv_hooks
+            .get_addr("delivery")
+            .unwrap_or_else(|| self.fin.clone());
+        let next_send_addr = self
+            .send_hooks
+            .get_addr("delivery")
+            .unwrap_or_else(|| self.fin.clone());
+
+        // Then we simply add a single SystemHandler stage
+        self.recv_hooks
+            .add(
+                Address::random(0),
+                "ordering",
+                hooks::ReceiverOrdering::default(),
+            )
+            .default(next_recv_addr);
+        self.send_hooks
+            .add(
+                Address::random(0),
+                "ordering",
+                hooks::SenderOrdering::default(),
+            )
+            .default(next_send_addr);
+        self
+    }
+
+    /// Enable the delivery guarantee behaviour on this pipe
+    ///
+    /// Additional behaviours can be added to compose a custom pipe
```

Comment:
Is there a `add_cusom_behaviour` API somewhere? Do we want this pipe to have a custom behaviour?

Reply:
So far there isn't, but fundamentally nothing is stopping us from allowing custom behaviour on pipes.

Reply:
Where would the custom behaviour go in the pipeline?
Since we control order of "ordering" and "redelivery" behaviours.

Reply:
I changed the initialisation of the pipe to do all the work in `build` for now. But I suspect we will want custom hooks on the sender side to go last, and on the receiver side to be first.
We could also have two functions to add hooks (`at_start` and `at_end` or something). But it's not needed for no and I wanna get this PR merged so let's leave it out for now